### PR TITLE
Update Nemo Distributed Checkpoint User Guide

### PR DESCRIPTION
**Pull request by FortunaZhang**
# What does this PR do ?

This PR aims to update Nemo distributed checkpoint user guide documentation, with best practices for parameter tuning, mechanism, FAQs, etc. 

**Collection**: [Note which collection this PR will affect]
Distributed checkpoint.

# Changelog 
- Update dist_ckpt.rst documentation in Nemo GitHub
- Update Introduction
- Add Mechanism with images
- Add Parameter Tuning Best Practices
- Add FAQs
- Add Glossary

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
- [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [x] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

### CODE REVIEW

File: docs/source/checkpoints/dist_ckpt.rst [hunk 1]
```diff
@@ -1,20 +1,125 @@
...
 Megatron Core provides a checkpointing library capable of handling all types of parallelisms used in LLM training.
 Although the distributed checkpointing library is targeted at the Megatron Core model, it can also be used with other models, as long as proper integration is implemented.
 
 The library provides two main entrypoints: ``dist_checkpointing.save`` and ``dist_checkpointing.load`` which are meant to replace the ``torch.save`` and ``torch.load`` in the regular checkpointing flow.
 Apart from that, it provides a mechanism to define how different types of local tensors should be combined and split in the global checkpoint.
 
 
+Mechanism
+--------------
+The NeMo Distributed Checkpoint enables saving and loading models from multiple ranks in parallel. It employs a novel strategy **Fully Parallel Saving** (FPS) to partition the optimizer states, gradients and model parameters across all GPU ranks. When saving the checkpoint of a distributed optimizer, each DP rank holds its shard of the optimizer state and independently writes its shard to the shared storage (grad buffer).
```
Comment (jgerh) [inline docs/source/checkpoints/dist_ckpt.rst:27]: fix punctuation, remove bold, revise

The NeMo Distributed Checkpoint enables saving and loading models from multiple ranks in parallel. It employs a novel strategy called Fully Parallel Saving (FPS) to partition the optimizer states, gradients, and model parameters across all GPU ranks. When saving the checkpoint of a distributed optimizer, each DP rank holds its shard of the optimizer state and independently writes its shard to the shared storage (grad buffer).

File: docs/source/checkpoints/dist_ckpt.rst [hunk 2]
```diff
@@ -1,20 +1,125 @@
...
+
+When loading the checkpoint, each DP rank reads its corresponding checkpoint file (shard) to recover. If different parallelism strategies are needed (e.g., tensor parallelism, pipeline parallelism), each rank can also access other checkpoint files to transfer data to the correct locations. 
+
+NeMo enables users to resume training from a checkpoint saved with different tensor and pipeline parallelism degrees, providing the flexibility to change training configurations as needed during training.
```
Comment (jgerh) [inline docs/source/checkpoints/dist_ckpt.rst:31]: suggested revision

NeMo allows users to resume training from a checkpoint saved with different tensor and pipeline parallelism degrees, offering the flexibility to adjust training configurations as needed.

File: docs/source/checkpoints/dist_ckpt.rst [hunk 3]
```diff
@@ -1,20 +1,125 @@
...
+.. image:: imgs/dist_ckpt_1.png
+
+
+*Figure 2. Asynchronous saving in Megatron Core saves checkpoint at the background in parallel with training*
+
+
+Parameter Tuning
+--------------
+
+Distributed checkpoints in NeMo could be configured for pre-training and fine-tuning jobs.
```
Comment (jgerh) [inline docs/source/checkpoints/dist_ckpt.rst:51]: revise to active voice, second person

You can configure distributed checkpoints in NeMo pre-training and fine-tuning jobs.

File: docs/source/checkpoints/dist_ckpt.rst [hunk 4]
```diff
@@ -1,20 +1,125 @@
...
+        	dist_ckpt_torch_dist_multiproc: 2
+
+        	dist_ckpt_assume_constant_structure: False
+
+        	dist_ckpt_parallel_dist_opt: True
+
+        	dist_ckpt_load_strictness: null
+
+
+Here are more details of the checkpoint format options and related parameters:
```
Comment (jgerh) [inline docs/source/checkpoints/dist_ckpt.rst:84]: revise

Here's a summary of the checkpoint format options and related parameters:

File: docs/source/checkpoints/dist_ckpt.rst [hunk 5]
```diff
@@ -425,3 +530,68 @@ and using the ``dist_checkpointing.save`` and ``dist_checkpointing.load`` entryp
...
+
+Distributed Optimizer
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+The distributed optimizer is a memory-optimized data-parallel deployment method. It shards the optimizer states and the high-precision master parameters across data-parallel GPUs instead of replicating them. At the parameter optimizer step, each data-parallel GPU updates its shard of parameters. Since each GPU needs its own gradient shard, the distributed optimizer conducts reduce-scatter of the parameter gradients instead of all-reduce of them. Then, the updated parameter shards are all-gathered across data-parallel GPUs. This approach significantly reduces the memory need of large-scale LLM training. Also, when the precision of the gradient is higher than the parameter precision, the split execution of gradient reduce-scatter and parameter all-gather can reduce the total communication volume. This split collective execution increases the total computation to overlap with the communication, which improves the overlap opportunity.
+
+More information please refer to https://docs.nvidia.com/nemo-framework/user-guide/latest/nemotoolkit/features/parallelisms.html.
```
Comment (jgerh) [inline docs/source/checkpoints/dist_ckpt.rst:597]: revise

For more information, please refer to https://docs.nvidia.com/nemo-framework/user-guide/latest/nemotoolkit/features/parallelisms.html.

**Review by jgerh — COMMENTED** [review]
Completed the technical edit of docs/source/checkpoints/dist_ckpt.rst and provided some copyedits and suggested revisions.

**Comment by FortunaZhang** [comment]
Updated per above suggestions, and changed the image source, please check again

**Review by jgerh — APPROVED** [review]
Approve revisions

**Review by yaoyu-33 — APPROVED** [review]
